File: models/Invoice.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Invoice = sequelize.define('Invoice', {
    id: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4
    },
    visit_id: {
        type: DataTypes.UUID,
        allowNull: false
    },
    institution_id: {
        type: DataTypes.UUID,
        allowNull: false
    },
    invoice_number: {
        type: DataTypes.STRING,
        allowNull: false,
        unique: true
    },
    invoice_date: {
        type: DataTypes.DATE,
        allowNull: false,
        defaultValue: DataTypes.NOW
    },
    due_date: {
        type: DataTypes.DATE,
        allowNull: false,
        defaultValue: () => new Date(Date.now() + 30 * 24 * 60 * 60 * 1000) // 30 days from now
    },
    subtotal: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0
    },
    tax_amount: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0
    },
    discount_amount: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0
    },
    total_amount: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0
    },
    amount_paid: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0
    },
    balance_due: {
        type: DataTypes.FLOAT,
        allowNull: false,
        defaultValue: 0,
        set(val) {
            this.setDataValue('balance_due', Math.round((this.total_amount - this.amount_paid) * 100) / 100);
        }
    },
    status: {
        type: DataTypes.ENUM('draft', 'unpaid', 'partially_paid', 'paid', 'cancelled', 'refunded'),
        defaultValue: 'draft'
    },
    payment_method: {
        type: DataTypes.ENUM('cash', 'credit_card', 'insurance', 'bank_transfer', 'mobile_money', 'other'),
        allowNull: true
    },
    notes: {
        type: DataTypes.TEXT,
        allowNull: true
    },
    metadata: {
        type: DataTypes.JSON,
        allowNull: true,
        defaultValue: {},
        comment: 'Additional invoice metadata like insurance details, billing codes, etc.'
    },
    created_by: {
        type: DataTypes.UUID,
        allowNull: true
    },
    patient_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    token: {
        type: DataTypes.STRING,
        allowNull: true,
        comment: 'Shareable token for public invoice viewing'
    },
    sms_sent: {
        type: DataTypes.BOOLEAN,
        allowNull: false,
        defaultValue: false
    },
    sms_sent_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    paid_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    paid_by: {
        type: DataTypes.UUID,
        allowNull: true
    },
    viewed_count: {
        type: DataTypes.INTEGER,
        allowNull: false,
        defaultValue: 0
    },
    viewed_at: {
        type: DataTypes.DATE,
        allowNull: true
    }
}, {
    tableName: 'invoices',
    timestamps: true,
    underscored: true,
    indexes: [
        { fields: ['token'], unique: true }
    ],
    hooks: {
        beforeSave: (invoice) => {
            // Always recalculate balance from source-of-truth amounts
            invoice.balance_due = Math.round((invoice.total_amount - invoice.amount_paid) * 100) / 100;

            // Auto-derive status so it can never drift out of sync with amounts.
            // Skip if already cancelled/refunded — those are manual state transitions
            // that should not be overridden by amount arithmetic.
            const skipStatus = ['cancelled', 'refunded'];
            if (!skipStatus.includes(invoice.status)) {
                if (invoice.balance_due <= 0 && parseFloat(invoice.amount_paid) > 0) {
                    invoice.status = 'paid';
                } else if (parseFloat(invoice.amount_paid) > 0 && invoice.balance_due > 0) {
                    invoice.status = 'partially_paid';
                } else {
                    invoice.status = 'unpaid';
                }
            }
        }
    }
});

// Associations
Invoice.associate = (models) => {
    Invoice.belongsTo(models.Visit, { foreignKey: 'visit_id', as: 'visit' });
    Invoice.belongsTo(models.Institution, { foreignKey: 'institution_id', as: 'institution' });
    Invoice.belongsTo(models.Staff, { foreignKey: 'created_by', as: 'creator' });
    // paid_by is a plain UUID — no FK constraint so both staff and admin IDs are accepted
    // Invoice.belongsTo(models.Staff, { foreignKey: 'paid_by', as: 'paidBy' });
    Invoice.belongsTo(models.Patient, { foreignKey: 'patient_id', as: 'patient' });
    Invoice.hasMany(models.ServiceBill, {
        foreignKey: 'invoice_id',
        as: 'service_bills',
    });

};

module.exports = Invoice;

File: billing-service/app/models/invoice.py
"""Invoice model - financial document per patient visit.
Aligned with the monolith's invoices table schema."""

import uuid
from datetime import datetime

from sqlalchemy import (
    Column, DateTime, Enum, Numeric, String, Text, Boolean,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base
from app.utils.types import InvoiceStatus


class Invoice(Base):
    __tablename__ = "invoices"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    invoice_number = Column(String(255), unique=True, nullable=False, index=True)

    # References (matching monolith schema exactly)
    patient_id = Column(UUID(as_uuid=True), nullable=True, index=True)
    visit_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    institution_id = Column(UUID(as_uuid=True), nullable=False, index=True)

    # Dates
    invoice_date = Column(DateTime(timezone=True), nullable=True)
    due_date = Column(DateTime(timezone=True), nullable=True)

    # Financial — all double precision in DB, use Numeric for Python
    subtotal = Column(Numeric(15, 2), default=0, nullable=False)
    tax_amount = Column(Numeric(15, 2), default=0, nullable=False)
    discount_amount = Column(Numeric(15, 2), default=0, nullable=False)
    total_amount = Column(Numeric(15, 2), default=0, nullable=False)
    amount_paid = Column(Numeric(15, 2), default=0, nullable=False)
    balance_due = Column(Numeric(15, 2), default=0, nullable=False)

    # Status
    status = Column(String(50), default="draft", index=True)
    payment_method = Column(String(50), nullable=True)

    # Metadata
    notes = Column(Text, nullable=True)
    metadata_ = Column("metadata", Text, nullable=True)

    # Audit
    created_by = Column(UUID(as_uuid=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)

    # Invoice sharing
    token = Column(String(255), nullable=True)
    sms_sent = Column(Boolean, default=False, nullable=False)
    sms_sent_at = Column(DateTime(timezone=True), nullable=True)
    paid_at = Column(DateTime(timezone=True), nullable=True)
    paid_by = Column(UUID(as_uuid=True), nullable=True)
    viewed_count = Column(Numeric, default=0, nullable=False)
    viewed_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    service_bills = relationship("ServiceBill", back_populates="invoice", lazy="selectin")
    payments = relationship("Payment", back_populates="invoice", lazy="selectin")
    nhia_claims = relationship("NHIAClaim", back_populates="invoice")

    def compute_totals(self):
        """Recalculate totals from service bills and payments."""
        self.subtotal = sum(float(b.total_amount or 0) for b in self.service_bills)
        self.total_amount = float(self.subtotal) + float(self.tax_amount or 0) - float(self.discount_amount or 0)
        self.total_amount = round(self.total_amount, 2)
        self.amount_paid = sum(float(p.amount or 0) for p in self.payments if p.status == "completed")
        self.balance_due = max(0, self.total_amount - self.amount_paid)
        self.balance_due = round(self.balance_due, 2)
